### DISCUSSION TITLE
Unable open for write,when reopen same file on Windows

### DISCUSSION THREAD

**Discussion by vicobill** [libvips/libvips#3013] [Q&A]
My code is very simple,in Python,use pyvips,on Windows.

```python
import os 
vipshome = r'c:\libvips\lib'
os.environ['PATH'] = vipshome + ';' + os.environ['PATH']

import pyvips

with pyvips.Image.new_from_file('some-image.jpg',access='sequential') as image:
    image.write_to_file('x.webp')

with pyvips.Image.new_from_file('x.webp',access='sequential') as image:
    image.write_to_file('y.jpg')

with pyvips.Image.new_from_file('another.jpg',access='sequential') as image:
    image.write_to_file('x.webp')  # this is the error occured. Error is 'unable to open for write'
```
I tried same code in macOS, it's ok, but in Windows, error occured, so weird.

Is the file handle not closed in Windows or anything else?  

libvips:8.13.0
pyvips:2.2.1

**Comment by vicobill** [comment]
```python
import os 
vipshome = r'c:\libvips\lib'
os.environ['PATH'] = vipshome + ';' + os.environ['PATH']

import pyvips

with pyvips.Image.new_from_file('some-image.jpg',access='sequential') as image:
    image.write_to_file('x.jpg')

with pyvips.Image.new_from_file('x.jpg',access='sequential') as image:
    image.write_to_file('y.jpg')

with pyvips.Image.new_from_file('another.jpg',access='sequential') as image:
    image.write_to_file('x.jpg')  
```
When save to .jpg,and reopen, things all right!

Does `WebPFree` not called?

**Comment by jcupitt** [comment]
Hi @vicobill,

I think this is because of the way libwebp load works -- you have to map the whole of the compressed image into memory, so that means you can't close the input file until after you call `WebPFree()`. Most readers (like JPG, for example) don't make you map the whole of the input.

libvips will cache image load operations, so it can't free until the webp loader leaves cache.
